src/main/java/com/amazonaws/encryptionsdk/keyrings/KmsKeyring.java

```diff
@@ -0,0 +1,171 @@
+        // MUST attempt to generate a new plaintext data key and encrypt that data key by calling KMS GenerateDataKey.
+        if (!encryptionMaterials.hasPlaintextDataKey()) {
+            generateDataKey(encryptionMaterials);
+        } else {
+            // If this keyring's generator is defined and was not used to generate a data key, OnEncrypt
+            // MUST also attempt to encrypt the plaintext data key using the CMK specified by the generator.
+            keyIdsToEncrypt.add(generatorKeyId);
+        }
+
+        // Given a plaintext data key in the encryption materials, OnEncrypt MUST attempt
+        // to encrypt the plaintext data key using each CMK specified in it's key IDs list.
+        for (String keyId : keyIdsToEncrypt) {
+            encryptDataKey(keyId, encryptionMaterials);
+        }
+    }
+
+    private void generateDataKey(final EncryptionMaterials encryptionMaterials) {
+        final GenerateDataKeyResult result = dataKeyEncryptionDao.generateDataKey(generatorKeyId,
+                encryptionMaterials.getAlgorithmSuite(), encryptionMaterials.getEncryptionContext());
+
+        encryptionMaterials.setPlaintextDataKey(result.getPlaintextDataKey(),
+                new KeyringTraceEntry(KMS_PROVIDER_ID, generatorKeyId, KeyringTraceFlag.GENERATED_DATA_KEY));
+        encryptionMaterials.addEncryptedDataKey(result.getEncryptedDataKey(),
+                new KeyringTraceEntry(KMS_PROVIDER_ID, generatorKeyId, KeyringTraceFlag.ENCRYPTED_DATA_KEY, KeyringTraceFlag.SIGNED_ENCRYPTION_CONTEXT));
+    }
+
+    private void encryptDataKey(final String keyId, final EncryptionMaterials encryptionMaterials) {
+        final EncryptedDataKey encryptedDataKey = dataKeyEncryptionDao.encryptDataKey(keyId,
+                encryptionMaterials.getPlaintextDataKey(), encryptionMaterials.getEncryptionContext());
+
+        encryptionMaterials.addEncryptedDataKey(encryptedDataKey,
+                new KeyringTraceEntry(KMS_PROVIDER_ID, keyId, KeyringTraceFlag.ENCRYPTED_DATA_KEY, KeyringTraceFlag.SIGNED_ENCRYPTION_CONTEXT));
+    }
+
+    @Override
+    public void onDecrypt(DecryptionMaterials decryptionMaterials, List<? extends EncryptedDataKey> encryptedDataKeys) {
+        requireNonNull(decryptionMaterials, "decryptionMaterials are required");
+        requireNonNull(encryptedDataKeys, "encryptedDataKeys are required");
+
+        if (decryptionMaterials.hasPlaintextDataKey() || encryptedDataKeys.isEmpty()) {
+            return;
+        }
+
+        if (!encryptedDataKeys.stream()
+                .filter(edk -> edk.getProviderId().equals(KMS_PROVIDER_ID))
+                .map(edk -> new String(edk.getProviderInformation(), PROVIDER_ENCODING))
+                .allMatch(KmsUtils::isArnWellFormed)) {
+            throw new MalformedArnException("encryptedDataKeys contains a malformed ARN");
+        }
+
+        for (EncryptedDataKey encryptedDataKey : encryptedDataKeys) {
```

Comment:
This is a potential O(n^2) operation. We iterate over each EDK and then in `okToDecrypt` iterate over every keyId we have configured. This seems like a potentially inefficient way to do this.

Reply:
I'll put all the configured keys into a HashSet to improve performance